File: baba-chatbot-python/src/baba_chatbot/integrations/llm/knowledge_repository.py
"""
Repositorio de conocimiento (RAG simple basado en archivos markdown).
Carga y busca información relevante en documentos de la base de conocimiento.
"""
import os
import logging
from typing import Dict, List, Tuple
from pathlib import Path


logger = logging.getLogger(__name__)


class KnowledgeRepository:
    """
    Repositorio de conocimiento que implementa RAG simple basado en archivos markdown.
    """

    def __init__(self, knowledge_base_path: str = None):
        """
        Inicializa el repositorio de conocimiento.
        
        Args:
            knowledge_base_path: Ruta al directorio con documentos markdown
        """
        if knowledge_base_path is None:
            knowledge_base_path = os.path.join(
                os.getcwd(), "..", "..", "config", "rag", "kb_sources"
            )
        
        self._knowledge_base_path = knowledge_base_path
        self._documents: Dict[str, str] = {}
        self._load_documents()

    def _load_documents(self) -> None:
        """
        Carga todos los documentos markdown del directorio de conocimiento.
        Omite archivos README.
        """
        try:
            logger.info(f"Loading knowledge base from: {self._knowledge_base_path}")

            if not os.path.exists(self._knowledge_base_path):
                logger.warning(f"Knowledge base directory not found: {self._knowledge_base_path}")
                return

            # Buscar todos los archivos markdown
            for root, _, files in os.walk(self._knowledge_base_path):
                for file in files:
                    if file.endswith('.md'):
                        file_name = Path(file).name
                        
                        # Omitir archivos README
                        if file_name.upper().startswith('README'):
                            continue

                        file_path = os.path.join(root, file)
                        try:
                            with open(file_path, 'r', encoding='utf-8') as f:
                                content = f.read()
                                self._documents[file_name] = content
                                logger.info(
                                    f"Loaded document: {file_name} ({len(content)} characters)"
                                )
                        except Exception as ex:
                            logger.error(f"Error reading file {file_path}: {ex}")

            logger.info(f"Knowledge base loaded: {len(self._documents)} documents")
        except Exception as ex:
            logger.error(f"Error loading knowledge base: {ex}")

    def search_relevant_context(self, query: str, max_chunks: int = 3) -> str:
        """
        Busca información relevante en la base de conocimiento usando búsqueda por palabras clave.
        Divide documentos en secciones y puntúa según relevancia.
        
        Args:
            query: Consulta del usuario
            max_chunks: Número máximo de secciones a retornar
            
        Returns:
            Contexto relevante concatenado
        """
        if not self._documents:
            logger.warning("No documents loaded in knowledge base")
            return ""

        try:
            # Extraer palabras clave de la consulta
            query_words = set(
                word.lower()
                for word in query.replace(',', ' ').replace('.', ' ')
                .replace('?', ' ').replace('!', ' ').split()
                if len(word) > 3
            )

            if not query_words:
                return ""

            # Buscar secciones relevantes
            relevant_sections: List[Tuple[str, str, int]] = []

            for doc_name, content in self._documents.items():
                sections = self._split_into_sections(content)
                
                for section in sections:
                    section_lower = section.lower()
                    score = sum(1 for word in query_words if word in section_lower)
                    
                    if score > 0:
                        relevant_sections.append((doc_name, section, score))

            # Ordenar por relevancia y tomar las mejores
            top_sections = sorted(
                relevant_sections,
                key=lambda x: x[2],
                reverse=True
            )[:max_chunks]

            if not top_sections:
                logger.debug(f"No relevant context found for query: {query}")
                return ""

            # Concatenar secciones
            context = "\n\n---\n\n".join(section for _, section, _ in top_sections)
            logger.info(f"Found {len(top_sections)} relevant sections for query")
            
            return context
        except Exception as ex:
            logger.error(f"Error searching knowledge base: {ex}")
            return ""

    def _split_into_sections(self, content: str) -> List[str]:
        """
        Divide un documento en secciones basándose en headers markdown.
        
        Args:
            content: Contenido del documento
            
        Returns:
            Lista de secciones
        """
        sections = []
        lines = content.split('\n')
        current_section = []

        for line in lines:
            # Detectar inicio de nueva sección
            if line.startswith('## ') or line.startswith('### '):
                if current_section:
                    section_text = '\n'.join(current_section).strip()
                    if len(section_text) > 50:
                        sections.append(section_text)
                    current_section = []

            current_section.append(line)

        # Agregar última sección
        if current_section:
            section_text = '\n'.join(current_section).strip()
            if len(section_text) > 50:
                sections.append(section_text)

        return sections

    def reload(self) -> None:
        """Recarga los documentos de la base de conocimiento."""
        self._documents.clear()
        self._load_documents()

    def get_stats(self) -> Tuple[int, int]:
        """
        Obtiene estadísticas de la base de conocimiento.
        
        Returns:
            Tupla con (número de documentos, total de caracteres)
        """
        total_chars = sum(len(doc) for doc in self._documents.values())
        return len(self._documents), total_chars

File: baba-chatbot-net/src/Baba.Chatbot.Integrations/Llm/KnowledgeRepository.cs
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.Logging;

namespace Baba.Chatbot.Integrations.Llm;

/// <summary>
/// Repositorio de conocimiento (RAG simple basado en archivos markdown)
/// Carga y busca información relevante en documentos de la base de conocimiento
/// </summary>
public class KnowledgeRepository
{
    private readonly ILogger<KnowledgeRepository> _logger;
    private readonly string _knowledgeBasePath;
    private readonly Dictionary<string, string> _documents = new();

    /// <summary>
    /// Inicializa una nueva instancia del repositorio de conocimiento
    /// </summary>
    public KnowledgeRepository(ILogger<KnowledgeRepository> logger, IConfiguration configuration)
    {
        _logger = logger;
        _knowledgeBasePath = configuration["KnowledgeBase:Path"] ?? 
                            Path.Combine(Directory.GetCurrentDirectory(), "..", "..", "config", "rag", "kb_sources");
        
        LoadDocuments();
    }

    /// <summary>
    /// Carga todos los documentos markdown del directorio de conocimiento
    /// Omite archivos README
    /// </summary>
    private void LoadDocuments()
    {
        try
        {
            var fullPath = _knowledgeBasePath;
            _logger.LogInformation("Loading knowledge base from: {Path}", fullPath);

            if (!Directory.Exists(fullPath))
            {
                _logger.LogWarning("Knowledge base directory not found: {Path}", fullPath);
                return;
            }

            var markdownFiles = Directory.GetFiles(fullPath, "*.md", SearchOption.AllDirectories);
            
            foreach (var file in markdownFiles)
            {
                var fileName = Path.GetFileName(file);
                if (fileName.StartsWith("README", StringComparison.OrdinalIgnoreCase))
                {
                    continue;
                }

                var content = File.ReadAllText(file);
                _documents[fileName] = content;
                _logger.LogInformation("Loaded document: {FileName} ({Size} characters)", fileName, content.Length);
            }

            _logger.LogInformation("Knowledge base loaded: {Count} documents", _documents.Count);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error loading knowledge base");
        }
    }

    /// <summary>
    /// Busca información relevante en la base de conocimiento usando búsqueda por palabras clave
    /// Divide documentos en secciones y puntúa según relevancia
    /// </summary>
    public string SearchRelevantContext(string query, int maxChunks = 3)
    {
        if (_documents.Count == 0)
        {
            _logger.LogWarning("No documents loaded in knowledge base");
            return string.Empty;
        }

        try
        {
            var queryWords = query.ToLower()
                .Split(new[] { ' ', ',', '.', '?', '!' }, StringSplitOptions.RemoveEmptyEntries)
                .Where(w => w.Length > 3)
                .ToHashSet();

            if (!queryWords.Any())
            {
                return string.Empty;
            }

            // Buscar documentos relevantes
            var relevantSections = new List<(string docName, string section, int score)>();

            foreach (var doc in _documents)
            {
                var sections = SplitIntoSections(doc.Value);
                
                foreach (var section in sections)
                {
                    var sectionLower = section.ToLower();
                    var score = queryWords.Count(word => sectionLower.Contains(word));
                    
                    if (score > 0)
                    {
                        relevantSections.Add((doc.Key, section, score));
                    }
                }
            }

            var topSections = relevantSections
                .OrderByDescending(s => s.score)
                .Take(maxChunks)
                .Select(s => s.section)
                .ToList();

            if (!topSections.Any())
            {
                _logger.LogDebug("No relevant context found for query: {Query}", query);
                return string.Empty;
            }

            var context = string.Join("\n\n---\n\n", topSections);
            _logger.LogInformation("Found {Count} relevant sections for query", topSections.Count);
            
            return context;
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error searching knowledge base");
            return string.Empty;
        }
    }

    /// <summary>
    /// Divide un documento en secciones basándose en headers markdown
    /// </summary>
    private List<string> SplitIntoSections(string content)
    {
        var sections = new List<string>();
        var lines = content.Split('\n');
        var currentSection = new List<string>();

        foreach (var line in lines)
        {
            if (line.StartsWith("## ") || line.StartsWith("### "))
            {
                if (currentSection.Any())
                {
                    var sectionText = string.Join('\n', currentSection).Trim();
                    if (sectionText.Length > 50)
                    {
                        sections.Add(sectionText);
                    }
                    currentSection.Clear();
                }
            }

            currentSection.Add(line);
        }

        if (currentSection.Any())
        {
            var sectionText = string.Join('\n', currentSection).Trim();
            if (sectionText.Length > 50)
            {
                sections.Add(sectionText);
            }
        }

        return sections;
    }

    /// <summary>
    /// Recarga los documentos de la base de conocimiento
    /// </summary>
    public void Reload()
    {
        _documents.Clear();
        LoadDocuments();
    }

    /// <summary>
    /// Obtiene estadísticas de la base de conocimiento
    /// </summary>
    public (int documentCount, int totalCharacters) GetStats()
    {
        var totalChars = _documents.Values.Sum(d => d.Length);
        return (_documents.Count, totalChars);
    }
}
